serverless/lambda/src/main/java/uk/gov/di/lambdas/LogoutHandler.java

```diff
@@ -4,29 +4,164 @@
+        }
+        if (!queryStringParameters.containsKey("id_token_hint")
+                || queryStringParameters.get("id_token_hint").isBlank()) {
+            sessionService.deleteSessionFromRedis(session.getSessionId());
+            return generateDefaultLogoutResponse(state);
+        }
+        if (!doesIDTokenExistInSession(queryStringParameters.get("id_token_hint"), session)) {
+            throw new RuntimeException(
+                    format("ID Token does not exist for Session: %s", session.getSessionId()));
+        }
+        if (!isIDTokenSignatureValid(
+                queryStringParameters.get("id_token_hint"), session.getSessionId())) {
+            throw new RuntimeException(
+                    format(
+                            "Unable to validate ID token signature for Session: %s",
+                            session.getSessionId()));
+        }
+        try {
+            String idTokenHint = queryStringParameters.get("id_token_hint");
+            SignedJWT idToken = SignedJWT.parse(idTokenHint);
+            Optional<String> audience =
+                    idToken.getJWTClaimsSet().getAudience().stream().findFirst();
+            sessionService.deleteSessionFromRedis(session.getSessionId());
+            return audience.map(
+                            t -> {
+                                final ClientRegistry clientRegistry;
+                                try {
+                                    clientRegistry =
+                                            dynamoClientService
+                                                    .getClient(t)
+                                                    .orElseThrow(
+                                                            () -> new ClientNotFoundException(t));
+                                } catch (ClientNotFoundException e) {
+                                    throw new RuntimeException(
+                                            format(
+                                                    "Client not found in ClientRegistry for ClientID: %s",
+                                                    t));
+                                }
+                                String logoutURI =
+                                        validateClientRedirectUri(
+                                                queryStringParameters, clientRegistry);
+                                return generateLogoutResponse(URI.create(logoutURI), state);
+                            })
+                    .orElse(generateDefaultLogoutResponse(state));
+        } catch (ParseException e) {
+            throw new RuntimeException();
+        }
+    }
+
+    private boolean isIDTokenSignatureValid(String idTokenHint, String sessionID) {
+        return true;
+    }
+
+    private boolean doesIDTokenExistInSession(String idTokenHint, Session session) {
+        for (Map.Entry<String, ClientSession> t : session.getClientSessions().entrySet()) {
```

Comment:
As we only use the values off the Entry, rather than both key and value, perhaps could be refactored into something like `sessions.getClientSessions().values().stream().[...].anyMatch(..)`

Reply:
Yep that is a lot neater. I have updated.